protocols/gossipsub/src/handler.rs

```diff
@@ -272,49 +277,65 @@ where
                     }
                 }
                 Some(OutboundSubstreamState::PendingSend(mut substream, message)) => {
-                    match substream.start_send(message)? {
-                        AsyncSink::Ready => {
-                            self.outbound_substream =
-                                Some(OutboundSubstreamState::PendingFlush(substream))
+                    match Sink::poll_ready(Pin::new(&mut substream), cx) {
+                        Poll::Ready(Ok(())) => {
+                            match Sink::start_send(Pin::new(&mut substream), message) {
+                                Ok(()) => {
+                                    self.outbound_substream =
+                                        Some(OutboundSubstreamState::PendingFlush(substream))
+                                }
+                                Err(e) => {
+                                    return Poll::Ready(ProtocolsHandlerEvent::Close(e));
+                                }
+                            }
+                        }
+                        Poll::Ready(Err(e)) => {
+                            // TODO: Is this what we want?
+                            return Poll::Ready(ProtocolsHandlerEvent::Close(e));
                         }
-                        AsyncSink::NotReady(message) => {
+                        Poll::Pending => {
                             self.outbound_substream =
                                 Some(OutboundSubstreamState::PendingSend(substream, message));
                             break;
                         }
                     }
                 }
                 Some(OutboundSubstreamState::PendingFlush(mut substream)) => {
-                    match substream.poll_complete()? {
-                        Async::Ready(()) => {
+                    match Sink::poll_flush(Pin::new(&mut substream), cx) {
+                        Poll::Ready(Ok(())) => {
                             self.outbound_substream =
                                 Some(OutboundSubstreamState::WaitingOutput(substream))
                         }
-                        Async::NotReady => {
+                        Poll::Ready(Err(e)) => {
+                            // TODO: Is this what we want?
+                            return Poll::Ready(ProtocolsHandlerEvent::Close(e))
```

Comment:
Originally I had intended this to remain connected if possible. If a send fails, my original thinking was that we could always re-open another outgoing substream. However if it fails here, there is probably some deeper underlying issue and we would want to end the connection. I think this is fine, and we can modify later if testing says otherwise.